src/controllers/handlers/ws-handler.ts

```diff
@@ -26,19 +27,148 @@ export async function registerWsHandler(
     Object.assign(data, { ...data, ...newData })
   }
 
+  function cleanupConnection(data: WsUserData, code: number, messageText: string) {
+    const { wsConnectionId } = data
+
+    if (isAuthenticated(data)) {
+      try {
+        data.transport.close()
+        rpcServer.detachUser(data.address)
+        data.eventEmitter.emit('close', { code, reason: messageText })
+        data.eventEmitter.all.clear()
+      } catch (error: any) {
+        logger.error('Error during connection cleanup', {
+          error: error.message,
+          address: getAddress(data),
+          wsConnectionId
+        })
+      }
+    }
+
+    if (isNotAuthenticated(data) && data.timeout) {
+      clearTimeout(data.timeout)
+      delete data.timeout
+    }
+
+    if (wsConnectionId) {
+      wsPool.releaseConnection(wsConnectionId)
+    }
+
+    changeStage(data, {
+      auth: false,
+      isConnected: false,
+      authenticating: false
+    })
+  }
+
+  function trackConnectionDuration(data: WsUserData) {
+    const { wsConnectionId, connectionStartTime } = data
+    const duration = (Date.now() - connectionStartTime) / 1000
+
+    if (!isNaN(duration) && isFinite(duration)) {
+      metrics.observe('ws_connection_duration_seconds', {}, duration)
+    }
+
+    logger.debug('WebSocket connection tracked duration', {
+      wsConnectionId: wsConnectionId,
+      durationSeconds: duration || 'N/A'
+    })
+  }
+
+  async function authenticateUser(ws: WebSocket<WsUserData>, data: WsNotAuthenticatedUserData, message: ArrayBuffer) {
+    try {
+      changeStage(data, { authenticating: true })
+
+      const authChainMessage = textDecoder.decode(message)
+
+      const verifyResult = await verify('get', '/', JSON.parse(authChainMessage), {
+        fetcher
+      })
+      const address = normalizeAddress(verifyResult.auth)
+
+      logger.debug('Authenticated User', { address, wsConnectionId: data.wsConnectionId })
+
+      const eventEmitter = mitt<IUWebSocketEventMap>()
+      const transport = await createUWebSocketTransport(ws, eventEmitter, config, logs)
+
+      changeStage(data, {
+        auth: true,
+        address,
+        eventEmitter,
+        isConnected: true,
+        transport,
+        authenticating: false
+      })
```

Comment:
Although this state change will work perfectly, what if we move it down as the last thing to do in the scope so no-one is tempted to put an async call after it which will switch to run something else concurrently and leave the `rpcServer` calls and callbacks without being set properly?

Reply:
Moved it to the `finally` clause. Good catch!